### PR TITLE
Add benchmark suite

### PR DESCRIPTION
**Pull request by lapp0**
- [x] benchmark `outlines/fsm/regex.py` numba function compilation
- [x] benchmark `outlines/fsm/json_schema.py` json to regex (`build_regex_from_object`)
- [x] benchmark `outlines/fsm/regex.py` to FSM
- [x] Document usage

No benchmark coverage for generation. Have some code written to do so, but would like this merged as a baseline before continuing.

### `pytest --benchmark-only --benchmark-columns=mean,max`

```
------------------------------------------ benchmark: 14 tests -------------------------------------------
Name (time in us)                                                 Mean                       Max          
----------------------------------------------------------------------------------------------------------
test_benchmark_json_schema_to_regex[simple_schema]             59.3640 (1.0)            172.8980 (1.49)   
test_benchmark_json_schema_to_regex[complex_schema]            59.3866 (1.00)           116.3650 (1.0)    
test_benchmark_regex_to_fsm[ip]                            59,500.5427 (>1000.0)     66,531.2210 (571.75) 
test_benchmark_regex_to_fsm[time]                          59,145.1800 (996.31)      65,897.8390 (566.30) 
test_benchmark_regex_to_fsm[date]                          60,256.4941 (>1000.0)     63,790.9300 (548.20) 
test_benchmark_regex_to_fsm[simple_phone]                  60,757.2710 (>1000.0)     63,508.6560 (545.77) 
test_benchmark_regex_to_fsm[ssn]                           63,519.3974 (>1000.0)     77,955.7490 (669.92) 
test_benchmark_regex_to_fsm[complex_phone]                 64,853.8790 (>1000.0)     86,702.4480 (745.09) 
test_benchmark_regex_to_fsm[email]                         64,185.6854 (>1000.0)     70,712.7130 (607.68) 
test_benchmark_regex_to_fsm[url]                           76,687.0116 (>1000.0)     85,416.9130 (734.04) 
test_benchmark_json_schema_to_fsm[simple_schema]           92,248.5355 (>1000.0)    105,986.8340 (910.81) 
test_benchmark_regex_to_fsm[quite_complex]                 90,073.5474 (>1000.0)     94,483.2810 (811.96) 
test_benchmark_json_schema_to_fsm[complex_schema]          96,999.0614 (>1000.0)    102,476.1180 (880.64) 
test_benchmark_compile_numba                            6,087,309.6160 (>1000.0)  9,463,025.1640 (>1000.0)
----------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

### Generate initial results:
```
git checkout benchmark-test-suite
pip install .
pytest -s tests/benchmark/test_benchmark_regex_fsm.py --benchmark-only --benchmark-columns=mean,max --benchmark-json=main-profile_regex_fsm.json
pytest -s tests/benchmark/test_benchmark_numba_compile.py --benchmark-only --benchmark-columns=mean,max --benchmark-json=main-profile_compile_numba.json

git checkout fsm-with-trie
pip install .
pytest -s tests/benchmark/test_benchmark_regex_fsm.py --benchmark-only --benchmark-columns=mean,max --benchmark-json=trie-profile_regex_fsm.json
pytest -s tests/benchmark/test_benchmark_numba_compile.py --benchmark-only --benchmark-columns=mean,max --benchmark-json=trie-profile_compile_numba.json
```

### Generate comparisons:

`py.test-benchmark compare --sort=fullname --columns=mean,max main-profile_regex_fsm.json trie-profile_regex_fsm.json`

```
-------------------------------------- benchmark: 18 tests ---------------------------------------
Name (time in s)                                                 Mean                Max          
--------------------------------------------------------------------------------------------------
test_benchmark_regex_to_fsm[complex_phone] (main-profile)      1.2403 (7.45)      1.4116 (8.14)   
test_benchmark_regex_to_fsm[complex_phone] (trie-profile)      0.3848 (2.31)      0.4245 (2.45)   
test_benchmark_regex_to_fsm[date] (main-profile)               0.7324 (4.40)      0.7493 (4.32)   
test_benchmark_regex_to_fsm[date] (trie-profile)               0.2405 (1.44)      0.3268 (1.88)   
test_benchmark_regex_to_fsm[email] (main-profile)              1.7789 (10.69)    10.3718 (59.78)  
test_benchmark_regex_to_fsm[email] (trie-profile)              2.2854 (13.73)    13.0081 (74.98)  
test_benchmark_regex_to_fsm[ip] (main-profile)                 0.5345 (3.21)      0.5558 (3.20)   
test_benchmark_regex_to_fsm[ip] (trie-profile)                 0.1739 (1.04)      0.1956 (1.13)   
test_benchmark_regex_to_fsm[quite_complex] (main-profile)     10.0635 (60.46)    10.1402 (58.45)  
test_benchmark_regex_to_fsm[quite_complex] (trie-profile)      3.8243 (22.98)     3.9292 (22.65)  
test_benchmark_regex_to_fsm[simple_phone] (main-profile)       0.3994 (2.40)      0.4185 (2.41)   
test_benchmark_regex_to_fsm[simple_phone] (trie-profile)       0.2163 (1.30)      0.2294 (1.32)   
test_benchmark_regex_to_fsm[ssn] (main-profile)                0.2920 (1.75)      0.3024 (1.74)   
test_benchmark_regex_to_fsm[ssn] (trie-profile)                0.1737 (1.04)      0.1788 (1.03)   
test_benchmark_regex_to_fsm[time] (main-profile)               0.2660 (1.60)      0.2833 (1.63)   
test_benchmark_regex_to_fsm[time] (trie-profile)               0.1664 (1.0)       0.1735 (1.0)    
test_benchmark_regex_to_fsm[url] (main-profile)                0.7149 (4.29)      0.7527 (4.34)   
test_benchmark_regex_to_fsm[url] (trie-profile)                0.9758 (5.86)      1.0571 (6.09)   
--------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

`py.test-benchmark compare --sort=fullname --columns=mean,max main-profile_compile_numba.json trie-profile_compile_numba.json`

```
-------------------------------- benchmark: 2 tests -------------------------------
Name (time in s)                                  Mean                Max          
-----------------------------------------------------------------------------------
test_benchmark_compile_numba (main-profile)     6.3623 (1.0)       9.9192 (1.0)    
test_benchmark_compile_numba (trie-profile)     8.4004 (1.32)     13.4900 (1.36)   
-----------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

### CODE REVIEW

File: tests/benchmark/test_benchmark_regex_fsm.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    "ip": r"(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)",
+    "url": r"(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?",
+    "ssn": r"\d{3}-\d{2}-\d{4}",
+    "quite_complex": "(['\"\\ ,]?((?:of|resulting|case|which|cultures|a|core|extreme|selflessness|spiritual|various|However|both|vary|in|other|secular|the|religious|among|moral|and|It|object|worldviews|altruism|traditional|material|aspect|or|life|beings|virtue|is|however|opposite|concern|an|practice|it|for|s|quality|religions|In|Altruism|animals|happiness|many|become|principle|human|selfishness|may|synonym)['\"\\ ,]?)+['\"\\ ,]?\\s\\|\\s([^|\\(\\)\n]{1,})\\s\\|\\s['\"\\ ,]?((?:of|resulting|case|which|cultures|a|core|extreme|selflessness|spiritual|various|However|both|vary|in|other|secular|the|religious|among|moral|and|It|object|worldviews|altruism|traditional|material|aspect|or|life|beings|virtue|is|however|opposite|concern|an|practice|it|for|s|quality|religions|In|Altruism|animals|happiness|many|become|principle|human|selfishness|may|synonym)['\"\\ ,]?)+['\"\\ ,]?(\\s\\|\\s\\(([^|\\(\\)\n]{1,})\\s\\|\\s([^|\\(\\)\n]{1,})\\))*\\n)*END\\ RELATIONS",
```
Comment (rlouf) [inline tests/benchmark/test_benchmark_regex_fsm.py:18]: Could we give this a more explicit name?

File: tests/benchmark/test_benchmark_regex_fsm.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+
+@pytest.mark.parametrize("regex_name", regex_samples.keys())
+def test_benchmark_regex_to_fsm(
+    benchmark, tokenizer, ensure_numba_compiled, regex_name
+):
+    """Benchmark converting regex to FSM"""
+    regex_str = regex_samples[regex_name]
+    benchmark.pedantic(
```
Comment (rlouf) [inline tests/benchmark/test_benchmark_regex_fsm.py:28]: Is there any reason why you are using the pedantic mode here?

**Comment by rlouf** [comment]
Thank you for opening a PR! Don't you think it would be best to always benchmark the end-to-end index computation?

File: tests/benchmark/test_benchmark_regex_fsm.py (hunk 2 above)
Comment (lapp0) [reply tests/benchmark/test_benchmark_regex_fsm.py:28]: IMO, it's cleaner than 

```
create_rfsm = lambda: RegexFSM(regex_str, tokenizer)
benchmark(create_rfsm)
```

File: tests/benchmark/test_benchmark_regex_fsm.py (hunk 1 above)
Comment (lapp0) [reply tests/benchmark/test_benchmark_regex_fsm.py]: Resolved with https://github.com/outlines-dev/outlines/pull/542/commits/ea1a0df40495a3c64848e224c862145df158a287

**Comment by lapp0** [comment]
> Thank you for opening a PR! Don't you think it would be best to always benchmark the end-to-end index computation? This is the quantity we care about.

Could you please clarify? We are benchmarking the computation of the `RegexFSM` index in this PR, which includes `create_fsm_index_end_to_end`. Which component would you like me to include?

**Comment by rlouf** [comment]
I mean not separating the Numba compilation from the rest of the index compilation. Total time is what we care about. Does that make sense?

**Comment by lapp0** [comment]
Numba initial compilation is a one time occurrence and takes ~9,400 ms per the posted benchmark. After compilation generating a `RegexFSM` takes on average between 59 and 97 ms. I think it makes sense to separate compilation from execution benchmarks.

**Comment by rlouf** [comment]
I can get on board with that. Would you mind fixing the merge conflicts?

**Comment by rlouf** [comment]
Thank you!

**Pull request merged**